Added rubocop.yml and corrected simple style warnings

iainbeeston opened the pull request:
I'd like to enable [HoundCI](https://houndci.com/) for json-schema. It works using [rubocop](https://github.com/bbatsov/rubocop) to warn of style violations on pull requests. Before we can do that, we need to establish a rubocop style definition for the project and correct existing style violations.

This PR adds rubocop and houndci config, and corrects inconsistencies in the code style used throughout the project. Where two different styles were in use (in different files) I've tried to use whichever style is most common throughout the codebase.

RST-J commented:
I also vote for adding rubocop as style checker. We use that in our company, too.
As a reference we use the [ruby styleguide ](https://github.com/bbatsov/ruby-style-guide).

I haven't read the whole config yet but I remember that rubocop by default has very strict cops regarding method and class size and complexity. So what exactly do want to have as limits in this respect? I mean as a goal, by now we won't score any acceptable values I guess.

iainbeeston commented:
The rubocop config I've added uses the default Hound CI as the base, and then I've customised it to the norms in this project.

All of the complexity and class/method length metrics are disabled at the moment. I've corrected every (reasonable) rubocop error, so that in this branch, there are no rubocop warnings at all.

RST-J commented:
Ok, I'll try to find time to investigate the config some time soon - I think we'll go for it, the question is if everyone agrees about the defined style.
The one thing that immediately came to my mind - no spaces in hash literals - is currently in my favour :grinning:

iainbeeston commented:
With the style config, I've tried to make it enforce the most common style throughout the codebase. For example, we use double-quotes more than single, so that became the enforced style. So hopefully it's representative, and just makes things more consistent. However I'd be more than happy to adjust it to suit tastes, or make it more strict

RST-J commented:
I don't want to provoke hard discussions, probably its all fine already or there are only minor things, I don't know.
But I think at least we four, i.e. also @hoxworth and @pd, should agree upon the config when we introduce it.

iainbeeston commented:
Sure, that sounds sensible. I'm not in a rush.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+    collect!: map!
+    find: detect
+    find_all: select
+    reduce: inject
```
I'd prefer reduce over inject here.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/GuardClause:
+  Description: Check for conditionals that can be replaced with guard clauses
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals
+  Enabled: false
```
I vote for enabling this.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/IfUnlessModifier:
+  Description: Favor modifier if/unless usage when you have a single-line body.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#if-as-a-modifier
+  Enabled: false
```
Here I also vote for true.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/SignalException:
+  Description: Checks for proper usage of fail and raise.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#fail-method
+  Enabled: false
```
I'd enable this. I like the differentiation between `fail` for an intended error and `raise` for something that comes from external problems.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/SingleLineMethods:
+  Description: Avoid single-line methods.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#no-single-line-methods
+  Enabled: false
```
Vote true.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/Alias:
+  Description: Use alias_method instead of alias.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#alias-method
+  Enabled: false
```
Vote true.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/OneLineConditional:
+  Description: Favor the ternary operator(?:) over if/then/else/end constructs.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#ternary-operator
+  Enabled: false
```
Vote true.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/VariableInterpolation:
+  Description: "Don't interpolate global, instance and class variables directly in strings."
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#curlies-interpolate
+  Enabled: false
```
Vote true.

RST-J commented on .rubocop.yml:
```diff
@@ -0,0 +1,253 @@
+Style/WordArray:
+  Description: Use %w or %W for arrays of words.
+  StyleGuide: https://github.com/bbatsov/ruby-style-guide#percent-w
+  Enabled: false
```
Vote true.

RST-J commented:
Do we want to go on with this? I'd say yes. The other two would have had enough time to react, I'd say.

bolshakov commented:
It should be safe to close this PR, rubocop was added here https://github.com/voxpupuli/json-schema/pull/480
